design-proposals/dynamic-primary-pod-nic-name.md

```diff
@@ -0,0 +1,466 @@
+
+1. The primary pod interface name is also required by virt-handler
+2. The virt-handler does not have access to pod objects
+
+#### Pros
+
+#### Cons
+
+- Requires a new mechanism to have the information in virt-handler.
+- All virt-launcher pods will have to use the downward API mechanism instead of just the subset using it today
+
+## Specifying the primary pod interface name
+
+### Adding a cluster-wide preferred pod NIC name
+
+A new field will be added in KubeVirt's cluster-wide network configuration, which will specify the preferred pod
+interface name that should be bound to the VM interface destined for pod network:
+
+```go
+type NetworkConfiguration struct {
+...
+PreferredPodPrimaryInterfaceName string `json:"preferredPodPrimaryInterfaceName,omitempty"`
+...
+}
+```
+
+#### API Example
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+metadata:
+  name: kubevirt
+  namespace: kubevirt
+spec:
+  configuration:
+    network:
+      preferredPodPrimaryInterfaceName: special-nic
+```
+
+virt-handler and virt-launcher will be adjusted to prefer this value when trying to bind the primary network interface.
+They will both fallback to `eth0` in case they cannot find the preferred interface name.
+
+#### Pros
+
+- Does not require Multus to work
+
+#### Cons
+
+- Requires manual configuration
+- Error prone
+- Has cluster-wide granularity
+- Requires a change to the communication protocol between virt-handler and virt-launcher to pass the cluster-wide
+  setting to virt-launcher
+
+### Specifying the primary pod NIC name in advance
+
+When creating a VM with secondary networks, KubeVirt specifies the pod interfaces’ names for all secondary interfaces.
+It is currently not possible to specify the primary interface’s name when templating the virt-launcher pod, since there
+is no K8s API that enables it.
+
+The primary pod interface name is set by the container runtime (for example cri-o), before invoking the default CNI.
+
+#### Pros
+
+#### Cons
+
+- There is no K8s API to name the pod's primary NIC
+- There is no Multus API or config to name the pod's primary NIC
+
+## Network Binding Plugins
+
+A virtual machine's NIC could be bound to the pod's primary NIC using
+a [network binding plugin](https://kubevirt.io/user-guide/network/network_binding_plugins/).
+
+### CNI
+
+KubeVirt's VMI controller is responsible for specifying the proper [NetworkSelectionElement](https://github.com/k8snetworkplumbingwg/network-attachment-definition-client/blob/506cfdac925790adf2f56f27740d2e87eaf2c83c/pkg/apis/k8s.cni.cncf.io/v1/types.go#L140)
+objects that will cause Multus to invoke a network binding plugin's CNI.
+
+Since this specification happens before the pod is created, it is not possible for KubeVirt to know in advance what will
+the pod's primary NIC's name be.
+
+Binding plugins' authors could add a configuration field to specify a list of possible primary interface names:
```

Comment:
This is one option.
- Another option is for the CNI to access k-api and query the relevant pod, looking at the `network-status`, similarly to what is suggested for the virt-controller.
- A third option is to request Multus to expose the `network-status` result through an env variable, for the CNIs to consume. This could be a futuristic enhancement request on Multus once we see it as a need from other sources.

Reply:
The first option is not currently feasible since Multus adds the `network-status` annotation after all the CNIs had completed successfully, please see:

https://github.com/k8snetworkplumbingwg/multus-cni/blob/bc6c8d5c76c775c3e8088631e7660e8a927b62c6/pkg/multus/multus.go#L763